spec/controllers/criteria_controller_spec.rb

```diff
@@ -858,78 +859,56 @@
-        expect(assignment.get_criteria(:all, :flexible).find_by(name: 'cr60').ta_visible).to be true
-        expect(assignment.get_criteria(:all, :flexible).find_by(name: 'cr60').peer_visible).to be false
+        expect(assignment.criteria.map(&:name)).to include('cr100', 'cr60')
+        expect(assignment.criteria.where(type: 'CheckboxCriterion').find_by(name: 'cr100').ta_visible).to be true
+        expect(assignment.criteria.where(type: 'CheckboxCriterion').find_by(name: 'cr100').peer_visible).to be false
+        expect(assignment.criteria.where(type: 'FlexibleCriterion').find_by(name: 'cr60').ta_visible).to be true
+        expect(assignment.criteria.where(type: 'FlexibleCriterion').find_by(name: 'cr60').peer_visible).to be false
       end
 
       it 'creates criteria with rounded (up to first digit after decimal point) maximum mark' do
         post_as admin, :upload, params: { assignment_id: assignment.id, upload_file: round_max_mark_file }
-        expect(assignment.get_criteria(:all, :rubric).first.name).to eq('cr90')
+        expect(assignment.criteria.where(type: 'RubricCriterion').first.name).to eq('cr90')
 
-        expect(assignment.get_criteria(:all, :rubric).first.max_mark).to eq(4.6)
+        expect(assignment.criteria.where(type: 'RubricCriterion').first.max_mark).to eq(4.6)
       end
       it 'does not create criteria with format errors in entries' do
         post_as admin, :upload, params: { assignment_id: assignment.id, upload_file: invalid_mixed_file }
 
-        expect(assignment.get_criteria.map(&:name)).not_to include('cr40', 'cr50', 'cr70')
+        expect(assignment.criteria.map(&:name)).not_to include('cr40', 'cr50', 'cr70')
         expect(flash[:error].map { |f| extract_text f })
           .to eq([I18n.t('criteria.errors.invalid_format') + ' cr40, cr70, cr50'].map { |f| extract_text f })
       end
 
       it 'does not create criteria with an invalid mark' do
         post_as admin, :upload, params: { assignment_id: assignment.id, upload_file: invalid_mixed_file }
 
-        expect(assignment.get_criteria.map(&:name)).not_to include('cr40', 'cr50')
+        expect(assignment.criteria.map(&:name)).not_to include('cr40', 'cr50')
       end
 
       it 'does not create criteria that have both visibility options set to false' do
         post_as admin, :upload, params: { assignment_id: assignment.id, upload_file: invalid_mixed_file }
 
-        expect(assignment.get_criteria.map(&:name)).not_to include('cr70')
+        expect(assignment.criteria.map(&:name)).not_to include('cr70')
       end
 
       it 'does not create criteria that have unmatched keys / more keys than required' do
+        expect(assignment.criteria.where(type: 'RubricCriterion').length).to eq(1)
         post_as admin, :upload, params: { assignment_id: assignment.id, upload_file: partially_valid_file }
-
-        criteria = assignment.get_criteria(:all, :rubric).first
```

Comment:
It was because there are errors that are raised so `ActiveRecord::Rollback` always gets called so the new criteria is never created. Because of this we only get returned the criteria created in the `before each` call at the beginning of this context, so I edited the test to check that there is an error raised for a 'partially invalid file'.

Reply:
Okay that makes sense, thanks!